### DISCUSSION TITLE
WAS docker image - no network address?

### DISCUSSION THREAD

**Discussion by dslugPX** [toverainc/willow#312] [General]
just looked at willow again for the first time in a few months. 

Figured I would update to the new stuff and so started with WAS, figuring that was the right move...

_Edit to add: I only have access to Macs, I tried on two different M1 devices, if that matters._ 

Per the docs - I run: docker run --detach --name=willow-application-server --pull=always --network=host --restart=unless-stopped --volume=was-storage:/app/storage ghcr.io/toverainc/willow-application-server

This starts up no problem, but no network address gets assigned. In the logs I see:

023-10-14 07:20:33 [2023-10-14T14:20:33] INFO:     Uvicorn running on http://0.0.0.0:8502 (Press CTRL+C to quit)

and you can see here on the interface in docker that it's not exposing. 

<img width="1035" alt="Screenshot 2023-10-14 at 7 28 04 AM" src="https://github.com/toverainc/willow/assets/113297415/334189a5-5aa6-47a9-b21f-9b8a0d99d069">

I presume I'm missing some small obvious step, but I can't seem to figure out what it is.
Thoughts?

**Comment by nikito** [comment]
The quick start guide here explains how to find the address: https://heywillow.io/quick-start-guide/
😊

**Comment by nikito** [comment]
Also to add, you don't see ports assigned because it is in host mode so there is no need for explicit mapping. Additionally when it shows 0.0.0.0:8502 that means the container is listening on port 8502 for all addresses/interfaces exposed to the container. 🙂

**Reply by dslugPX** [reply]
Right, but it's not listening on either the hostname or ip, nor on any loopback I can find.

**Reply by sammcj** [reply]
Just a note that ideally you wouldn't ever run it (or any container) in host mode though.

**Comment by mikey60** [comment]
The Docker display you show looks like it is From Docker Desktop for Windows. If that is the case, Docker is most likely running under WSL2.  So you have to do some extra steps to map the WSL2 Docker ports to the host computer ports.  I am running on a Windows 10 machine and I assume it is the same on a Windows 11 machine.  I actually run a script that I found on the internet in a Windows PowerShell run with administrator rights to map the ports for WIS and WAS.  I did a quick search for a link to the web page where I found the approach I used, but could not find it.  There are many web pages that have information about mapping WSL2 docker ports to the host.

**Reply by dslugPX** [reply]
mac OS in this case...

**Reply by mikey60** [reply]
I know nothing about mac OS.

**Comment by eifinger** [comment]
I followed [this post from the homeassistant forum](https://community.home-assistant.io/t/willow-voice-assistant/572004/74):

Inside WSL2:

```console
ip route get 1.1.1.1 | grep -oP 'src \K\S+' 
```

In a Powershell with Admin rights:

```console
netsh interface portproxy add v4tov4 listenport=8502 listenaddress=0.0.0.0 connectport=8502 connectaddress=<IP_FROM_PREVIOUS_STEP>
```

Then open the port in the windows firewall.

**Comment by tersenius** [comment]
I had exactly the same issue. Problem is --network=host is not supported by Docker on macOS.

My solution was changing the network mode to bridge and publishing port 8502:
`docker run --detach --name=willow-application-server --pull=always --network=bridge -p 8502:8502 --restart=unless-stopped --volume=was-storage:/app/storage ghcr.io/toverainc/willow-application-server`

Then access to WAS was possible at  \<ip of macOS host\>:8502

Maybe this works as well in Windows, cannot test though.

**Reply by dslugPX** [reply]
Hey thanks so much for this! Was driving me completely bonkers!

I'm back up and running thanks to this (and H/T to the devs in the discord who mentioned this was solved here, I hadn't looked in a while)

**Reply by JeffCrum1** [reply]
This also works on Windows using WSL2.  No need for extra commands in WSL or Powershell.
